### ISSUE TITLE
Copy values extracted from topic match into payload?

### ISSUE THREAD

**Issue by DavidAntliff** [SpotlightKid/osc2mqtt#1]
Is it possible to take a value from an OSC topic (e.g. `/1/pattern/X` where X is 1,2,3,4) and somehow get this into the MQTT value?

I.e. if `/1/pattern/1` is received, I want to map this to value `1` sent to MQTT topic `app/pattern`. And if `/1/pattern/2` is received, I want to map this to value `2` sent to the same topic.

Although I could write individual rules for every possible value of X, this isn't going to work well in my case because I have 256 possible values of X, and that would require either a lot of MQTT subscriptions, or subscription to `app/#` and a lot of topic processing on the MQTT client.

**Comment by SpotlightKid** [comment]
Sorry, this is not supported by osc2mqtt atm. You'd have to change the `from_osc`method of `o2mqtt.convrter.Osc2MqttConverter` to do that. You could, for example, pass the group matches from the `match` regular expression as additional arguments to `encode_values`.

**Comment by SpotlightKid** [comment]
I've added a branch / pull request with a preliminary implementation of the requested feature in #2.

Please try it out and let me know if this handles your situation. I've added an example conversion rule section to the example configuration file (`osc2mqtt.ini`), which I think handles the scenario you described.

**Comment by DavidAntliff** [comment]
Thanks for this, I will try it out now and report back.

**Comment by DavidAntliff** [comment]
BTW, slightly unrelated, but I wanted to do a `pip install -e .` on the branch of your PR but there's an unresolved symlink called `parse_requirements.py` that prevents this. Not a big deal as I can run it in-place using the `start.sh` script, but I thought I'd mention it just in case.

**Comment by DavidAntliff** [comment]
Ok, my ignorance of OSC may be telling here. First, thank you for taking the time to implement this PR. Let me explain in a bit more about what I'm trying to do:

I have TouchOSC (on iPhone) and I want to implement a "mode selector" - imagine a 4x1 grid of mutually exclusive toggle buttons, such that only one of each button can be "on" at any time. In the end I would like to use an 8x8 grid of such buttons, but 4 will be fine for now.

I have an embedded application (on a WiFi connected microcontroller) that subscribes to MQTT topics. To keep memory requirements low, I want to limit the number of subscriptions (although subscribing to "#" seems OK). This app needs to know when a button becomes active.

TouchOSC sends these button updates through as OSC messages with the address `/page1/button1/X/Y` where `X` is the column index and `Y` is the row index. When a button goes active the message has _value_ 1.0, and when inactive 0.0. What I had hoped to do was have a single MQTT topic that receives the ID of the button that is now active as the MQTT message value, but I have just realised that this won't work in the case where there is more than one row, because I'll need to identify the row index `Y` also. Perhaps one way is to send both `X` and `Y` as string values, space delimited, as the MQTT value.

But the fact that two messages are sent for every button press (one to enable, one to disable), means I'm wondering now if my request even makes sense, because how would osc2mqtt know whether an incoming OSC message was an enable (1.0) or disable (0.0) for the addressed button? Perhaps all three values could be combined, somehow, so that the MQTT client receives a message on, say, "app/pattern" with a string of the format "button_x_id button_y_id state".

Can you see a way to handle this with the change you've made so far? If not, my apologies if I've wasted your time with this.

**Comment by DavidAntliff** [comment]
First attempt (I'm not at all confident that this is correct):

```
[:button]
match = ^/(?P<page>\w+)/(?P<button>\w+)/(?P<X>\d+)/(?P<Y>\d+)
address = /{page}/{button}/{_values[0]}
topic = {page}/{button}
address_groups = X, Y
from_osc = int
type = string
format = ascii
```

Produces:

```
$ ./start.sh 
INFO:Connecting to MQTT broker localhost:1883 ...
DEBUG:Starting MQTT thread...
INFO:Starting OSC server listening on port 9001 ...
DEBUG:MQTT connect: Connection Accepted.
DEBUG:OSC recv: /page1/button1/2/1 [0.0]
DEBUG:Rule 'button' match on: /page1/button1/2/1
Traceback (most recent call last):
  File "src/liblo.pyx", line 55, in liblo._weakref_method.__call__
  File "osc2mqtt/__init__.py", line 138, in handle_osc
    res = self.converter.from_osc(oscaddr, values, tags)
  File "osc2mqtt/converter.py", line 197, in from_osc
    values = values + extra_values
TypeError: can only concatenate list (not "tuple") to list
Exception TypeError: 'can only concatenate list (not "tuple") to list' in 'liblo._msg_callback' ignored
DEBUG:OSC recv: /page1/button1/3/1 [1.0]
DEBUG:Rule 'button' match on: /page1/button1/3/1
Traceback (most recent call last):
  File "src/liblo.pyx", line 55, in liblo._weakref_method.__call__
  File "osc2mqtt/__init__.py", line 138, in handle_osc
    res = self.converter.from_osc(oscaddr, values, tags)
  File "osc2mqtt/converter.py", line 197, in from_osc
    values = values + extra_values
TypeError: can only concatenate list (not "tuple") to list
Exception TypeError: 'can only concatenate list (not "tuple") to list' in 'liblo._msg_callback' ignored
```

**Comment by SpotlightKid** [comment]
> BTW, slightly unrelated, but I wanted to do a pip install -e . on the branch of your PR but there's an unresolved symlink called parse_requirements.py that prevents this.

Doh. Fixed this in the `master` branch and rebased the PR (i.e. this might mean you have to check out the branch again).

**Comment by SpotlightKid** [comment]
Fixed the exception when using more than one group in `address_groups` or `topic_groups`.

Your configuration is almost correct, except that you need two elements in the `from_osc` setting, one conversion function for each value added by `address_groups` and also, if you want the conversion in the MQTT->OSC direction to work, add both values back to the OSC address. In this case, also make sure that you have `^/?` add the start of the `match` setting value, to make the slash prefix optional and make the `/<X>/<Y>` part also optional.

Here's what I I just tested successfully (I'm using JSON format payload, because it's easier to read in the log output with option `-v`):

```
[:button]
match = ^/?(?P<page>\w+)/(?P<button>\w+)(/(?P<X>\d+)/(?P<Y>\d+))?
address = /{page}/{button}/{_values[0]}/{_values[1]}
topic = {page}/{button}
address_groups = X, Y
from_osc = int, int
type = json
format = ascii
```

**Comment by DavidAntliff** [comment]
Sorry, other things have come up but I will return to this as time allows.

**Comment by SpotlightKid** [comment]
That's ok. I didn't have the chance yet, to try your use case with TouchOSC either. Will try to do so after christmas.

**Comment by SpotlightKid** [comment]
This has been implemented in 8ea73bbd911aca520ea9102ad6817e472d0a86d2 now.

**Comment by DavidAntliff** [comment]
Thank you.
